pkg/controller/uniteddeployment/utils/ref_manager.go

```diff
@@ -0,0 +1,222 @@
+package utils
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// GetOwner is the type of function which is provided to get the owner resource.
+type GetOwner func() (runtime.Object, error)
+
+// UpdateOwnee is the type of function which is provided to update a resource after ownerReference is updated.
+type UpdateOwnee func(runtime.Object) error
+
+// RefManager provides the method to
+type RefManager struct {
+	getOwner    GetOwner
+	updateOwnee UpdateOwnee
+	selector    labels.Selector
+	owner       metav1.Object
+	ownerType   reflect.Type
+	schema      *runtime.Scheme
+
+	once        sync.Once
+	canAdoptErr error
+}
+
+// NewRefManager returns a RefManager that exposes
+// methods to manage the controllerRef of pods.
+func NewRefManager(getOwner GetOwner, updateOwnee UpdateOwnee, selector *metav1.LabelSelector, owner metav1.Object, schema *runtime.Scheme) (*RefManager, error) {
+	s, err := metav1.LabelSelectorAsSelector(selector)
+	if err != nil {
+		return nil, err
+	}
+
+	ownerType := reflect.TypeOf(owner)
+	if ownerType.Kind() == reflect.Ptr {
+		ownerType = ownerType.Elem()
+	}
+	return &RefManager{
+		getOwner:    getOwner,
+		updateOwnee: updateOwnee,
+		selector:    s,
+		owner:       owner,
+		ownerType:   ownerType,
+		schema:      schema,
+	}, nil
+}
+
+// ClaimOwnedObjects tries to take ownership of a list of objects for this controller.
+func (mgr *RefManager) ClaimOwnedObjects(objs []metav1.Object, filters ...func(metav1.Object) bool) ([]metav1.Object, error) {
+	match := func(obj metav1.Object) bool {
+		if !mgr.selector.Matches(labels.Set(obj.GetLabels())) {
+			return false
+		}
+
+		for _, filter := range filters {
+			if !filter(obj) {
+				return false
+			}
+		}
+
+		return true
+	}
+
+	claimObjs := []metav1.Object{}
+	errlist := []error{}
+	for _, obj := range objs {
+		ok, err := mgr.claimObject(obj, match)
+		if err != nil {
+			errlist = append(errlist, err)
+			continue
+		}
```

Comment:
remove `continue` and change to else if ok ?

Reply:
done